src/cdomains/intDomain.ml

```diff
@@ -911,30 +944,631 @@ struct
+
+  let bitand ik x y =
+    let interval_bitand = bit Ints_t.bitand ik in
+    binop x y interval_bitand
+
+  let bitor ik x y =
+    let interval_bitor = bit Ints_t.bitor ik in
+    binop x y interval_bitor
+
+  let bitxor ik x y =
+    let interval_bitxor = bit Ints_t.bitxor ik in
+    binop x y interval_bitxor
+
+  let bitnot ik x =
+    let interval_bitnot i =
+      match interval_to_int i with
+      | Some x -> of_int ik (Ints_t.bitnot x) |> fst
+      | _ -> top_of ik
+    in
+    unop x interval_bitnot
+
+  let shift_left ik x y =
+    let interval_shiftleft = bitcomp (fun x y -> Ints_t.shift_left x (Ints_t.to_int y)) ik in
+    binary_op_with_ovc x y interval_shiftleft
+
+  let shift_right ik x y =
+    let interval_shiftright = bitcomp (fun x y -> Ints_t.shift_right x (Ints_t.to_int y)) ik in
+    binary_op_with_ovc x y interval_shiftright
+
+  let lognot ik x =
+    let log1 f ik i1 =
+      match interval_to_bool i1 with
+      | Some x -> of_bool ik (f x)
+      | _ -> top_of ik
+    in
+    let interval_lognot = log1 not ik in
+    unop x interval_lognot
+
+  let logand ik x y =
+    let interval_logand = log (&&) ik in
+    binop x y interval_logand
+
+  let logor ik x y =
+    let interval_logor = log (||) ik in
+    binop x y interval_logor
+
+  let add ?no_ov ik x y =
+    let interval_add ((x1, x2), (y1, y2)) = [(x1 +. y1, x2 +. y2)] in
+    binary_op_with_norm ik x y interval_add
+
+  let neg ?no_ov ik x =
+    let neg_interval ((x, y)) = [(Ints_t.neg y, Ints_t.neg x)] in
+    unary_op_with_norm ik x neg_interval
+
+  let sub ?no_ov ik x y =
+    let interval_sub ((x1, x2), (y1, y2)) = [(x1 -. y2, x2 -. y1)] in
+    binary_op_with_norm ik x y interval_sub
+
+  let mul ?no_ov (ik: ikind) (x: t) (y: t) =
+    let interval_mul ((x1, x2), (y1, y2)) =
+      let x1y1 = x1 *. y1 in
+      let x1y2 = x1 *. y2 in
+      let x2y1 = x2 *. y1 in
+      let x2y2 = x2 *. y2 in
+      [(min4 x1y1 x1y2 x2y1 x2y2, max4 x1y1 x1y2 x2y1 x2y2)]
+    in
+    binary_op_with_norm ik x y interval_mul
+
+  let div ?no_ov ik x y =
+    let rec interval_div ((x1, x2), (y1, y2)) = begin
+      let is_zero v = v =. Ints_t.zero in
+      match y1, y2 with
+      | l, u when is_zero l && is_zero u -> top_of ik (* TODO warn about undefined behavior *)
+      | l, _ when is_zero l              -> interval_div ((x1,x2), (Ints_t.one,y2))
+      | _, u when is_zero u              -> interval_div ((x1,x2), (y1, Ints_t.(neg one)))
+      | _ when leq (of_int ik (Ints_t.zero) |> fst) ([(y1,y2)]) -> top_of ik
+      | _ ->
+        let x1y1n = x1 /. y1 in
+        let x1y2n = x1 /. y2 in
+        let x2y1n = x2 /. y1 in
+        let x2y2n = x2 /. y2 in
+        let x1y1p = x1 /. y1 in
+        let x1y2p = x1 /. y2 in
+        let x2y1p = x2 /. y1 in
+        let x2y2p = x2 /. y2 in
+        [(min4 x1y1n x1y2n x2y1n x2y2n, max4 x1y1p x1y2p x2y1p x2y2p)]
+    end
+    in binary_op_with_norm ik x y interval_div
```

Comment:
I'm surprised by the amount of `IntervalFunctor` code this just duplicates as opposed to delegating to the existing implementation. I thought it was supposed to just do that. Now we have two copies of the massive interval implementation to keep in sync.

Reply:
I have now tried to get more reuse out of this by pulling out operations that are basically textbook interval arithmetic. However, I am not sure how much more we can really pull out here.

Reply:
Is there a reason interval sets cannot just delegate directly to our interval domain for all operations? I mean not just extracting some common internal arithmetic, but just delegate the whole operation.

Reply:
The problem is that  e.g. normalization is different between both: IntervalSets can represent intervals spanning over zero for unsigned types without going to top. This also means that the type is different, as normalization for IntervalSests would lead to intervals here (to avoid having intervals with an upper bound smaller than the lower bound).

Reply:
> IntervalSets can represent intervals spanning over zero for unsigned types without going to top.


So `IntervalSet` is not just a disjunctive completion of intervals, which simply lifts all operations directly, (as described by e.g. Miné's numeric tutorial) but additionally borrows elements from the circular intervals domain that we used to have for such wrap-around intervals?
It would be good to have such details documented because I always had the impression that this is just a lifting to sets (implemented as sorted lists) with a normalization procedure (which joins overlapping intervals).

Reply:
I added a comment to this effect now.

Reply:
030aebc780a815fcf294f3e659515dc4f04ef477 added the comment to the existing single interval domain, not the new interval set domain.

Reply:
True, wasn't paying close enough attention apparently.